File: backend/app/audio/chroma_extractor.py
"""
app/audio/chroma_extractor.py
──────────────────────────────
Converts raw PCM audio chunks into 12-dimensional chroma vectors in real-time.

Design choices for 2026
───────────────────────
• librosa.feature.chroma_cqt  →  uses a Constant-Q Transform which gives
  better pitch resolution at low frequencies than STFT-based chroma.

• Buffer management  →  incoming chunks are accumulated in a ring buffer
  so that each chroma frame is computed over a full N_FFT window while
  advancing by HOP_LENGTH, maintaining temporal continuity across packets.

• Latency estimate   →  with sr=22050 and hop=512, one frame ≈ 23 ms.
  A 4-frame window ≈ 93 ms, well within the 100 ms requirement.

• Noise gate         →  frames below -60 dBFS are zero-padded chroma
  to avoid noisy pitch estimates during silence.
"""
from __future__ import annotations

import numpy as np
import librosa
from collections import deque
from typing import Deque, List, Optional

from app.core.config import settings


class ChromaExtractor:
    """
    Stateful, streaming chroma extractor.

    Usage::

        ex = ChromaExtractor()
        # When audio chunk arrives (float32 PCM, any length):
        frames = ex.push(pcm_chunk)
        for chroma_12 in frames:
            position = dtw.step(chroma_12)
    """

    # Silence threshold: RMS below this → emit zero chroma
    NOISE_GATE_RMS: float = 10 ** (-60 / 20)   # −60 dBFS

    def __init__(
        self,
        sr: int = settings.SAMPLE_RATE,
        hop_length: int = settings.HOP_LENGTH,
        n_fft: int = settings.N_FFT,
        n_chroma: int = settings.N_CHROMA,
        mic_gain: float = 1.0,
    ) -> None:
        self.sr = sr
        self.hop = hop_length
        self.n_fft = n_fft
        self.n_chroma = n_chroma
        self.gain = mic_gain

        # Streaming carry buffer: holds the samples that have not yet been
        # consumed as a full frame.  Always shorter than n_fft.
        self._buf: np.ndarray = np.zeros(0, dtype=np.float32)

    # ── Public API ───────────────────────────────────────────────────────────────

    def push(self, pcm: bytes | np.ndarray, sensitivity: float = 1.0) -> List[np.ndarray]:
        """
        Ingest raw PCM data and return a list of chroma frames produced.

        Parameters
        ----------
        pcm         : bytes or float32 ndarray
            Raw audio samples.  Bytes assumed to be float32 little-endian.
        sensitivity : float
            Multiplied into the signal before extraction (user mic sensitivity).

        Returns
        -------
        List of chroma vectors (each shape (12,)).
        """
        if isinstance(pcm, (bytes, bytearray)):
            audio = np.frombuffer(pcm, dtype=np.float32).copy()
        else:
            audio = np.asarray(pcm, dtype=np.float32)

        audio *= self.gain * sensitivity

        # Prepend the samples carried over from the previous call, then frame the
        # combined signal on a single global grid (n_fft window, hop stride).
        # Framing this way means feeding a signal whole or split into many chunks
        # yields the *same* frames (no reprocessing of overlaps).
        combined = np.concatenate([self._buf, audio])

        n_frames = 0
        if len(combined) >= self.n_fft:
            n_frames = 1 + (len(combined) - self.n_fft) // self.hop

        if n_frames <= 0:
            # Not enough for a full window yet – keep buffering.
            self._buf = combined.astype(np.float32, copy=True)
            return []

        chromavecs = self._compute_chroma_batch(combined, n_frames)

        # Carry the not-yet-consumed remainder (< n_fft samples) forward.
        self._buf = combined[n_frames * self.hop:].astype(np.float32, copy=True)

        return chromavecs

    def reset(self) -> None:
        """Clear internal buffer (call when starting a new session)."""
        self._buf = np.zeros(0, dtype=np.float32)

    # ── Private ──────────────────────────────────────────────────────────────────

    def _compute_chroma_batch(
        self, combined: np.ndarray, n_frames: int
    ) -> List[np.ndarray]:
        """
        Compute all L2-normalised chroma frames for ``combined`` in one shot.

        Uses an STFT-based chroma (a single vectorised transform for the whole
        buffer) rather than a per-frame constant-Q transform.  A per-frame CQT
        with ``hop_length == n_fft`` costs ~300 ms/frame, roughly 14x slower than
        real time; the batched STFT is ~1 ms/frame and easily meets the latency
        budget.  Because the reference features are built synthetically from the
        score's pitch classes (not from audio), an STFT chroma is fully
        comparable to them under the follower's cosine distance.

        A noise gate zeroes frames quieter than ``NOISE_GATE_RMS`` so silence
        does not produce spurious pitch estimates.
        """
        # Chroma for every frame; center=False keeps framing on our global grid.
        chroma = librosa.feature.chroma_stft(
            y=combined,
            sr=self.sr,
            n_fft=self.n_fft,
            hop_length=self.hop,
            n_chroma=self.n_chroma,
            center=False,
        ).astype(np.float32)                       # shape (n_chroma, n_frames)

        # Per-frame RMS for the noise gate (same framing as the chroma).
        windows = librosa.util.frame(
            combined, frame_length=self.n_fft, hop_length=self.hop
        )                                          # shape (n_fft, n_frames)
        rms = np.sqrt(np.mean(windows ** 2, axis=0))

        out: List[np.ndarray] = []
        for j in range(min(n_frames, chroma.shape[1])):
            if rms[j] < self.NOISE_GATE_RMS:
                out.append(np.zeros(self.n_chroma, dtype=np.float32))
                continue
            c = chroma[:, j]
            norm = float(np.linalg.norm(c))
            if norm > 1e-7:
                c = c / norm
            out.append(c.astype(np.float32))
        return out

File: frontend/lib/offline/chroma_extractor.dart
// lib/offline/chroma_extractor.dart
// On-device streaming chroma extractor. Mirrors the backend's deterministic
// chroma (explicit STFT -> power -> chroma filterbank -> L2 norm, fixed tuning),
// validated bit-for-bit against the Python engine. Pure Dart.

import 'dart:math';
import 'dart:typed_data';

import 'fft.dart';

class ChromaExtractor {
  static const int nFft = 2048;
  static const int hop = 512;
  static const int nChroma = 12;
  static const int nBins = nFft ~/ 2 + 1; // 1025
  static final double _gate = pow(10, -60 / 20).toDouble(); // -60 dBFS

  /// Chroma filterbank, row-major (nChroma * nBins), same matrix as the backend
  /// (librosa.filters.chroma at tuning 0). Loaded once from a bundled asset.
  final Float32List filterbank;

  final FFT _fft = FFT(nFft);
  final Float64List _hann = Float64List(nFft);
  final Float64List _re = Float64List(nFft);
  final Float64List _im = Float64List(nFft);
  Float32List _carry = Float32List(0);

  ChromaExtractor(this.filterbank) {
    assert(filterbank.length == nChroma * nBins);
    for (int i = 0; i < nFft; i++) {
      _hann[i] = 0.5 - 0.5 * cos(2 * pi * i / nFft);
    }
  }

  void reset() => _carry = Float32List(0);

  /// Push float32 PCM (22050 Hz mono) and get the chroma frames produced.
  /// Framing runs on one global grid (carry the unconsumed remainder) so
  /// streaming yields the same frames as processing the signal whole.
  List<Float32List> push(Float32List samples) {
    final combined = Float32List(_carry.length + samples.length);
    combined.setAll(0, _carry);
    combined.setAll(_carry.length, samples);

    final total = combined.length;
    final nFrames = total < nFft ? 0 : 1 + ((total - nFft) ~/ hop);
    final out = <Float32List>[];
    for (int i = 0; i < nFrames; i++) {
      out.add(_frame(combined, i * hop));
    }
    final consumed = nFrames * hop;
    _carry = Float32List.sublistView(combined, consumed);
    return out;
  }

  Float32List _frame(Float32List y, int off) {
    double ss = 0;
    for (int k = 0; k < nFft; k++) {
      final s = y[off + k];
      ss += s * s;
    }
    final c = Float32List(nChroma);
    if (sqrt(ss / nFft) < _gate) return c; // noise gate -> zero vector

    for (int k = 0; k < nFft; k++) {
      _re[k] = y[off + k] * _hann[k];
      _im[k] = 0.0;
    }
    _fft.transform(_re, _im);

    for (int p = 0; p < nChroma; p++) {
      double acc = 0;
      final base = p * nBins;
      for (int b = 0; b < nBins; b++) {
        acc += filterbank[base + b] * (_re[b] * _re[b] + _im[b] * _im[b]);
      }
      c[p] = acc;
    }
    double nrm = 0;
    for (int p = 0; p < nChroma; p++) nrm += c[p] * c[p];
    nrm = sqrt(nrm);
    if (nrm > 1e-7) {
      for (int p = 0; p < nChroma; p++) c[p] /= nrm;
    }
    return c;
  }
}
